### ISSUE TITLE
GlyphLayout unexpected whitespace omitted

### ISSUE THREAD

**Issue by scientistnm** [libgdx/libgdx#3230]
Hello,

I will try to explain what I'm trying to do and what is the problem here.
If you don't consider this as a bug, then please explain how to overcome this behavior of GlyphLayout.

I'm working on the 3D framework for TV platforms, than uses Libgdx underneath it.
I'm trying to create 3D input field widget. It is basicallly a decal with texture that I've prepared in runtime while drawing BitmapFont into FBO.

I noticed a problem that GlyphLayout omitts multiple whitespaces, and then my text isn't rendering correctly.
Also it omitts white spaces in the end and start of the run.
This is the problem for me because when I try to calculate position of the cursor in the input field...
String that should be rendered and the count of Glyphs in GlyphLayout is not the same.
Some of whitespaces are omitted during wrap, and those two are not in sync anymore.

I sync the cursor index manually, RIGHT key (cursorIndex++), LEFT key (cursorIndex--).
So when I try to put cursor in the next line...the index of selected character in the string is not the same as the index of glyph in glyph layout...and I miss the correct position of where cursor should be.

I have a workaround for it, but I'm not sure if this is the right way to do...

In the GlyphLayout class in wrap method, when I do this everything looks good in my usecase:

// Determine index where first run ends, ignoring whitespace before the wrap index.
int endIndex = wrapIndex;
//for (; endIndex > 0; endIndex--)
//	if (!fontData.isWhitespace((char)first.glyphs.get(endIndex - 1).id)) break;

// Determine index where second run starts, ignoring whitespace after the wrap index.
     int startIndex = wrapIndex;
//for (; startIndex < glyphCount; startIndex++)
//if (!fontData.isWhitespace((char)first.glyphs.get(startIndex).id)) break;

So basically I prevented the white space omittment, and everything draws correct if I have multiple whitespaces together, also the number of characters in String and the number of glyphs in GlyphLayout is the same, and caluclation of cursor position is correct.
Is the the right way to achieve this?

If this is the only way...I can send a pull request that changes the visibility of this method from private to protected, and I can override this behaviour.

What do you think about this?

**Comment by NathanSweet** [comment]
Please show the problem with an SSCCE.

**Comment by scientistnm** [comment]
Here you go.
Either way string lenght is 25, if text is wrapped on white space, number of glyphs will be reduced, and not be in sync with the given string.
In this use case string lenght is 25, glyphCount is 20.
Is this correct behavior?

public class GlyphLayoutWhitespaceTest  extends GdxTest {

	
	private GlyphLayout layout;
	private String text = "0123456789     0123456789";
	private int width = 20;
	private BitmapFont font;
	
	@Override
	public void create () {
		
		font = new BitmapFont();
		layout = new GlyphLayout(font, text, Color.WHITE, width, Align.left, true);
		
		int glyphCount = 0;
		for(GlyphRun run : layout.runs){
				glyphCount+= run.glyphs.size;
		}
		
		System.out.println("STRING LENGHT " + text.length() + " GLYPHS " + glyphCount);
		
	}
	
}

**Comment by NathanSweet** [comment]
The number of glyphs doesn't need to match the text. This is the point of [shaping](http://www.badlogicgames.com/wordpress/?p=3658), which is converting characters into glyphs for rendering.

**Comment by scientistnm** [comment]
Ok, I understand that. Thats why I asked is there a way to make workaround that.
I have looked in the TextArea.java and TextField.java...but having troubles understand what is really going on...it is lacking out of comments.

Besides that, I noticed that when I type in my text field (I append new char to existing string, and ask bitmap font to draw it)...When the new line is wrapped on whitespace (new line is beginning with white space)...it will not draw the following multiple whitespaces.
I think that this may be a side effect of omitting multiple whitespaces.

It would be nice if you can give me some instructions and advices how to achieve this.
Basically I'm reinventing the wheel here, with the TextField.java but drawing it's content to texture and applying it to decal.

**Comment by NathanSweet** [comment]
What would it draw for whitespace? Whitespace is nothing.

You are writing a new TextField? Copy/paste TextField and make it do what you want, or use TextField as a reference. Each GlyphRun has `xAdvances`, see javadoc for that field and how TextField uses it.
